Worker restart on function timeout

yojagad opened the pull request:
Fixes #5391.

yojagad committed:
Initial commit

mhoeger commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -405,5 +405,15 @@ public void Dispose()
+        public async Task RestartAsync()
+        {
+            // Dispose and restart all initialized channels
+            var channels = await GetInitializedWorkerChannelsAsync();
```
I think you should consider adding
```
State = FunctionInvocationDispatcherState.WorkerProcessRestarting;
```

Since we're re-starting all of these processes, we don't want any new requests landing on these old workers. There is still a chance that a worker process restarts and sets state back to "Initialized" (line 125) before all workers are disposed (say we have 10), but this would be a start in preventing that.

pragnagopa replied:
+1 For adding new state:WorkerProcessRestarting

yojagad committed:
Addressed comment

yojagad commented on src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs:
```diff
@@ -48,8 +54,15 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
-
-            _applicationLifetime.StopApplication();
+            string workerRuntime = _environment.GetEnvironmentVariable(RpcWorkerConstants.FunctionWorkerRuntimeSettingName);
+            if (string.IsNullOrEmpty(workerRuntime) || string.Equals(workerRuntime, RpcWorkerConstants.DotNetLanguageWorkerName, StringComparison.InvariantCultureIgnoreCase))
```
@pragnagopa - Discussed this with @mhoeger and null could also mean out of proc in some scenarios, right? If yes, should there be a separate tracking issue where "FunctionWorkerRuntimeSettingName" should be the ultimate source of truth without ambiguities ?

pragnagopa replied:
Need to rely on `FunctionInvocationDispatcherstate`.  FunctionInvocationDispatcherstate is set to initialized only for out-of-proc.

pragnagopa replied:
Do not rely on 'workerRuntime' as for a given language it can change out-of-proc or not. For example, when we add a language worker for C#, this is not true.

yojagad replied:
Ah, yes. This would be useless if dotnet goes out-of-proc. Agreed.

mhoeger replied:
What if we look for initialized channels that are actively executing the function that timed out and just shut that one down? This relies on the fact that although WebJobs [knows that the function execution has failed](https://github.com/Azure/azure-webjobs-sdk/blob/f7da27d133b9ce0aadf9918770f6ed45748f3c8a/src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs#L291), the RpcWorkerChannel still believes that the invocation is an [executingInvocation](https://github.com/Azure/azure-functions-host/blob/7d449c87e6684cf5cbc2507ba189e3158c9d2ea6/src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs#L43) (maybe that's not a safe assumption in the long-run). And if nothing is found, shut down the whole application? This would help if/when we have multiple language worker processes running

yojagad replied:
@mhoeger I like that idea. Not entirely sure if this would be a much larger enhancement , hmm..

pragnagopa replied:
This just came up in a CRI analysis. App is using multiple worker processes - 10. Only invocation on one of them timeouted out, rest of the 9 processes are still healthy. What you and @mhoeger proposed makes more sense. It also, makes sense to only restart worker channels that are faulted due to timeout. So introduce, another state for channel 'Faulted' and only restart channels that are faulted.

pragnagopa commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -405,5 +405,16 @@ public void Dispose()
+        {
+            // Dispose and restart all initialized channels
+            State = FunctionInvocationDispatcherState.WorkerProcessRestarting;
+            var channels = await GetInitializedWorkerChannelsAsync();
```
You need to restart all channels not just the ones that are 'Initialized'

yojagad replied:
What would be a scenario where a channel is in "Initializing" or "Default" state when a timeout happens? If invocations are only happening through "Initialized" channels, doesn't it make sense to restart only those? Unless I'm missing something here...

pragnagopa replied:
A worker channel is initialized after  worker startes and grpc channel is established...not restarting all the channels will lead to a race condition if jobhost restart is issued before grpc channel is initialized.

mhoeger replied:
@pragnagopa - i think i'm also confused here. I don't think there will be a jobhost restart on a timeout? Or are you talking about a jobhost restart for a different reason?

pragnagopa replied:
@mhoeger - this fix is about restarting jobhost if function invocations timeout instead of shutting down whole application

mhoeger replied:
@pragnagopa - i think the change is only to restart the worker channels, whether they are at jobhost or webhost scope (which makes sense - because functions and host configuration are not excepted to change on a timeout).

yojagad replied:
The change is only around restarting worker channels (and therefore recycling the processes), and not an entire jobhost restart..

pragnagopa replied:
My bad. I recommend JobHost restart which aligns well with holding off reading events from event sources until jobhost restart completes. This would be cleaner approach.

pragnagopa replied:
Taking my comment back. Lets test the changes where restarting channels works :)

mhoeger replied:
I don't think that we wait for worker channels to be up and running to read events from event sources, so a JobHost restart seems like overkill? Although if we can change that in the future this wouldn't benefit from that change. (I ran a quick test to add a 20 second delay in sending worker init response to simulate slow startup:)
<img width="813" alt="screenshot" src="https://user-images.githubusercontent.com/8888531/76895422-3047be00-684d-11ea-8175-26f89aadefbd.png">

yojagad replied:
I tested this out and noticed the process id change after recycling.
![image](https://user-images.githubusercontent.com/26885664/76896236-aef12b00-684e-11ea-87af-a44a7463ca7e.png)

pragnagopa replied:
This just came up in a CRI analysis. App is using multiple worker processes - 10. Only invocation on one of them timeouted out, rest of the 9 processes are still healthy. What you and @mhoeger proposed makes more sense. It also, makes sense to only restart worker channels that are faulted due to timeout. So introduce, another state for channel 'Faulted' and only restart channels that are faulted.

pragnagopa commented on src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs:
```diff
@@ -48,8 +54,15 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
+            }
+            else
+            {
+                await _functionInvocationDispatcher.RestartAsync();
```
Add warning level log indicating Jobhost is restarting

yojagad replied:
Will do. But also, this is not a jobhost restart.. Just want to make sure we're on the same page here :)

mhoeger commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -405,5 +405,16 @@ public void Dispose()
+            var channels = await GetInitializedWorkerChannelsAsync();
+            foreach (var channel in channels)
+            {
+                await DisposeAndRestartWorkerChannel(_workerRuntime, channel.Id);
```
Maybe too big of a change for this PR, but we should probably also drain executions here before shutting down (because a timeout, unlike an unhandled exception on the process, could have other healthy executions running too). Maybe we file an issue as a separate improvement (application shutdown has graceful timeouts on executions, so we are giving some of that up).

yojagad replied:
We already have a grace period in the timeout exception itself though.
![image](https://user-images.githubusercontent.com/26885664/76896522-363e9e80-684f-11ea-8f18-153203a9f649.png)

mhoeger replied:
I think new requests are still going to be routed to the worker though? So even though there's the 2 second grace period, existing requests are finishing and new requests are just starting to execute?

yojagad replied:
I think that's the behavior today as well even with normal shutdown. Even if we give grace periods for execution to complete, as long as the web app itself is accepting requests, there are chances some additional executions might flow into the worker/channel shutting down..

mhoeger replied:
@yojagad  - wasn't the change you made to add [DrainInvocationsAsync](https://github.com/Azure/azure-functions-host/commit/4a9aa46e9442ea9d4b0b68fbe6d5c730b8c90baf) in order to complete outstanding executions, while not allowing new ones (because the jobhost is shut down at that point)? Or maybe we need to add `State = Stopped` or something to our StopAsync to make that happen.

but either way! since we're just recycling workers we should be able to do better i think..!

(but i am actually curious if we can get concretely down the ordering of events, especially as it relates to https://github.com/Azure/azure-functions-host/issues/5785 and potentially this https://github.com/Azure/azure-functions-host/issues/5651)

mhoeger commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -185,5 +185,10 @@ public Task ShutdownAsync()
+
+        public async Task RestartAsync()
+        {
+            await DisposeAndRestartWorkerChannel(_httpWorkerChannel.Id);
```
Add `State = FunctionInvocationDispatcherState.WorkerProcessRestarting;` here too

mhoeger commented on src/WebJobs.Script/Workers/Http/HttpFunctionInvocationDispatcher.cs:
```diff
@@ -185,5 +185,10 @@ public Task ShutdownAsync()
         }
+
+        public async Task RestartAsync()
+        {
```
Add `State = FunctionInvocationDispatcherState.WorkerProcessRestarting;` here too

yojagad replied:
State is already being changed as the first step within DisposeAndRestartWorkerChannel() which is why I didn't add it here... This was not the case for RpcFunctionInvocationDispatcher though.

mhoeger replied:
Ahh got it! I think it's in a different place for rpc because we only want to delay invocations if there are 0 workers alive. Sounds good, thanks!

yojagad committed:
Changed approach to restart only errored channels

yojagad commented:
Changed approach for RPC channels to dispose and restart only the errored channels. @mhoeger @pragnagopa If this approach looks alright, I'll go ahead and start adding tests to it. Ran the same test as I previously did and was able to see the channel go into error state and subsequently, a different process.

yojagad committed:
Disposing registration

yojagad commented:
Changed approach to restart only channels that are in error. @mhoeger, @pragnagopa  If this approach looks alright, I'll start adding tests to it.

pragnagopa commented on src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs:
```diff
@@ -39,17 +42,24 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
-            _applicationLifetime.StopApplication();
+            if (_functionInvocationDispatcher.State.Equals(FunctionInvocationDispatcherState.Initialized))
+            {
+                _logger.LogWarning("Restarting language worker processes due to function timing out.", exceptionInfo.SourceException);
```
Nit: Use similar wording `A function timeout has occurred. Restarting worker process.`

pragnagopa commented on src/WebJobs.Script/ScriptHostBuilderExtensions.cs:
```diff
@@ -196,6 +195,8 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
                 }
                 services.TryAddSingleton<FunctionsScaleManager>();
+
+                services.AddSingleton(s => s.GetRequiredService<IFunctionInvocationDispatcherFactory>().GetFunctionDispatcher());
```
Avoid adding this as a dependency. FunctionInvocationDispatcherFactory is available. Use that instead

pragnagopa commented on src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:
```diff
@@ -11,13 +11,18 @@ public enum RpcWorkerChannelState
+        /// <summary>
+        /// LanguageWorkerChannel is in error state. Worker process is Initialized. Rpc Channel is established.
+        /// </summary>
+        Errored,
```
Nit: Faulted?

yojagad replied:
Does not apply anymore.

pragnagopa commented on src/WebJobs.Script/Workers/Rpc/FunctionRegistration/RpcFunctionInvocationDispatcher.cs:
```diff
@@ -405,5 +405,16 @@ public void Dispose()
             Dispose(true);
         }
+
+        public async Task RestartAsync()
```
Rename method. RestartFaultedChannelsAsync or RestartErroredChannelsAsync

yojagad replied:
Not applicable anymore.

pragnagopa reviewed (commented):
Added minor comments. Approach looks good. Will do another round of reviews after unit tests are added. Thanks.

yojagad committed:
Addressed comments as discussed offline.

yojagad commented:
After discussion offline, changed the approach to take advantage of using invocation id that comes with the exception to figure out which worker to restart

mhoeger commented:
Draft looks good to me! This avoids any race condition between two separate timers and their callbacks.

pragnagopa commented:
@yojagad @mhoeger - Earlier commit which propagated cancellation token from invoker to rpcworkerchannel is a good approach as it reduces the gap between an invocation state between Host and Rpc layer. And it will allow future extensions when we need to start sending invocation cancel messages to worker process.

You also, need to ensure to change the state for worker channel to : Faulted to ensure new invocations are not scheduled to the worker channel that is about to be restarted.

yojagad committed:
Updated enum

pragnagopa commented on src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:
```diff
@@ -8,16 +8,16 @@ public enum RpcWorkerChannelState
         /// LanguageWorkerChannel is created. Worker process is Initialized. Rpc Channel is established.
         /// </summary>
-        Initialized,
+        Initialized = 1 << 2,
```
i just sent a PR updating this to Flags :) take a look!

yojagad committed:
Added Tests

yojagad committed:
Changes after rebase

yojagad committed:
Updated condition to account for channels being disposed.

yojagad committed:
Addressed CR comment

yojagad commented:
After discussion with @Fabio and @pragnagopa , it was agreed that passing down the cancellationToken need to be taken as a separate work item. So, sticking to the current approach for now. Filed a tracking work item here https://github.com/Azure/azure-functions-host/issues/5828

Just rebased the PR and added tests. Let me know if I missed anything , please.

mhoeger commented on src/WebJobs.Script.WebHost/WebScriptHostExceptionHandler.cs:
```diff
@@ -39,17 +42,30 @@ public async Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, T
+                    _logger.LogWarning($"Restarting all language worker processes since invocation Id '{timeoutException.InstanceId}' was not found.", exceptionInfo.SourceException);
+                    await functionInvocationDispatcher.RestartAllWorkersAsync();
+                }
+                _logger.LogWarning("Restart of language worker process(es) completed.", exceptionInfo.SourceException);
```
Should this be a warning or info?

yojagad replied:
One of the previous comments suggested warning..

mhoeger reviewed (dismissed):
✅

pragnagopa commented on src/WebJobs.Script/Workers/Rpc/RpcWorkerChannel.cs:
```diff
@@ -111,7 +111,7 @@ internal class RpcWorkerChannel : IRpcWorkerChannel, IDisposable
         public bool IsChannelReadyForInvocations()
         {
-            return _state.HasFlag(RpcWorkerChannelState.InvocationBuffersInitialized | RpcWorkerChannelState.Initialized);
+            return !_disposing && !_disposed && _state.HasFlag(RpcWorkerChannelState.InvocationBuffersInitialized | RpcWorkerChannelState.Initialized);
```
Add test where this method returns false if a restart is issued on the channel

yojagad replied:
Added a test where this returns false when it's being disposed.

pragnagopa commented on src/WebJobs.Script/Workers/Rpc/RpcWorkerChannelState.cs:
```diff
@@ -11,21 +11,21 @@ public enum RpcWorkerChannelState
         /// LanguageWorkerChannel is created. InvocationBuffers per function are setup
         /// </summary>
-        InvocationBuffersInitialized = 1,
+        InvocationBuffersInitialized = 1 << 1,
```
Any reason why this change is needed?

yojagad replied:
I was changing the Default's value ( 0 cannot be a flag value unless it's used to represent 'None' of the flags are set ) and also updated the rest just for readability sake.

pragnagopa replied:
instead of using bitwise operator, for simplicity can you change the values to : 1, 2, 4 ....?

mhoeger replied:
I kinda prefer the bitwise operator too - it makes it easier for then next person to code and not make a math mistake!

pragnagopa replied:
alritie..which ever makes it easier to extend this. Will resolve this comment.

pragnagopa commented on test/WebJobs.Script.Tests.Integration/Host/WebScriptHostManagerTimeoutTests.cs:
```diff
@@ -56,31 +56,49 @@ public async Task OnTimeoutException_UsesToken_ManagerKeepsRunning()
+        [Fact]
+        public async Task OnTimeoutException_OOP_HasExpectedLogs()
+        {
+            using (var host = await RunTimeoutExceptionTest(handleCancellation: false, argKey: "inputData", timeoutFunctionName: "TimeoutSync", path: @"TestScripts\Node"))
```
Also, add E2E test here: https://github.com/Azure/azure-functions-host/blob/dev/test/WebJobs.Script.Tests.Integration/WebHostEndToEnd/SamplesEndToEndTests_Node.cs ( or any language) that verifies process id is different for worker process and Hostinstance id which is same after the restart.

yojagad replied:
Added validation for node processes, how do I get the HostInstanceId from TestFunctionHost?

pragnagopa reviewed (commented):
Code changes look ok.  Added more comments on adding more tests. Thanks!

TurtleBeach commented:
@pragnagopa Don't give up so easily. 4 characters (1 << 1) is more cognitive load than 1, as is the math to know what is intended. I would argue most developers do not day in / day out twiddle bits, and so the << also takes a few extra milliseconds in the brain to process.  4 characters is also 4 chances for a typo, vs 1 out of sequence integer. 1, 2, 4 are explicit. Use them. My .02

yojagad committed:
Addressed code review comments and added tests

yojagad committed:
Updated test

yojagad committed:
Addressed CR comments

pragnagopa reviewed (approved):
👍

mhoeger reviewed (approved):
✅
